graph-proxy/src/graphql/workflows.rs

```diff
@@ -185,6 +189,114 @@ impl TryFrom<IoArgoprojWorkflowV1alpha1WorkflowStatus> for WorkflowErroredStatus
     }
 }
 
+#[allow(clippy::missing_docs_in_private_items)]
+#[derive(Debug, Enum, PartialEq, Eq, Clone, Copy)]
+enum TaskStatus {
+    Pending,
+    Running,
+    Succeeded,
+    Skipped,
+    Failed,
+    Error,
+    Omitted,
+}
+
+impl TryFrom<String> for TaskStatus {
+    type Error = WorkflowParsingError;
+
+    fn try_from(status: String) -> Result<Self, <Self as TryFrom<String>>::Error> {
+        match status.as_str() {
+            "Pending" => Ok(TaskStatus::Pending),
+            "Running" => Ok(TaskStatus::Running),
+            "Succeeded" => Ok(TaskStatus::Succeeded),
+            "Skipped" => Ok(TaskStatus::Skipped),
+            "Failed" => Ok(TaskStatus::Failed),
+            "Error" => Ok(TaskStatus::Error),
+            "Omitted" => Ok(TaskStatus::Omitted),
+            _ => Err(WorkflowParsingError::UnrecognisedPhase),
+        }
+    }
+}
+
+/// A Task created by a workflow
+#[derive(Debug, SimpleObject)]
+struct Task {
+    /// Unique name of the task
+    id: String,
+    /// Display name of the task
+    name: String,
+    /// Current status of a task
+    status: TaskStatus,
+    /// Parent of a task
+    depends: Vec<String>,
+}
+
+/// A wrapper for list of tasks
+struct TaskMap(HashMap<String, IoArgoprojWorkflowV1alpha1NodeStatus>);
+
+impl TryFrom<argo_workflows_openapi::IoArgoprojWorkflowV1alpha1NodeStatus> for Task {
+    type Error = WorkflowParsingError;
+
+    fn try_from(
+        value: argo_workflows_openapi::IoArgoprojWorkflowV1alpha1NodeStatus,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            id: value.id,
+            name: value
+                .display_name
+                .ok_or(WorkflowParsingError::UnrecognisedPhase)?
+                .to_string(),
+            status: TaskStatus::try_from(
+                value
+                    .phase
+                    .ok_or(WorkflowParsingError::UnrecognisedPhase)?
+                    .to_string(),
+            )?,
+            depends: vec![],
+        })
+    }
+}
+
+impl TaskMap {
+    /// Generates a relationship map between parent and children
+    fn generate_relationship_map(&self) -> HashMap<String, Vec<String>> {
+        let mut parent_map: HashMap<String, Vec<String>> = HashMap::new();
+
+        for (node_name, node_status) in &self.0 {
+            for child in &node_status.children {
+                parent_map
+                    .entry(child.clone())
+                    .or_default()
+                    .push(node_name.clone());
+            }
+        }
+        parent_map
+    }
+}
+
+impl TryFrom<TaskMap> for Vec<Task> {
+    type Error = WorkflowParsingError;
+
+    fn try_from(value: TaskMap) -> Result<Self, Self::Error> {
+        let relationship_map = value.generate_relationship_map();
+
+        let tasks = value
+            .0
+            .into_iter()
+            .map(|(node_name, node_status)| {
+                let mut task = Task::try_from(node_status)?;
```

Comment:
Instead of creating from the `node_status` then mutating with the dependencies, could we create a single function which takes both and produces a task?